system_zero/src/arithmetic/mod.rs

```diff
@@ -0,0 +1,60 @@
+use plonky2::field::extension_field::Extendable;
+use plonky2::field::field_types::Field;
+use plonky2::field::packed_field::PackedField;
+use plonky2::hash::hash_types::RichField;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use starky::vars::StarkEvaluationTargets;
+use starky::vars::StarkEvaluationVars;
+
+use crate::arithmetic::addition::{eval_addition, eval_addition_recursively, generate_addition};
+use crate::arithmetic::division::{eval_division, eval_division_recursively, generate_division};
+use crate::arithmetic::multiplication::{
+    eval_multiplication, eval_multiplication_recursively, generate_multiplication,
+};
+use crate::arithmetic::subtraction::{
+    eval_subtraction, eval_subtraction_recursively, generate_subtraction,
+};
+use crate::column_layout::arithmetic::*;
+use crate::column_layout::NUM_COLUMNS;
+use crate::public_input_layout::NUM_PUBLIC_INPUTS;
+
+mod addition;
+mod division;
+mod multiplication;
+mod subtraction;
+
+pub(crate) fn generate_arithmetic_unit<F: RichField>(values: &mut [F; NUM_COLUMNS]) {
+    if values[IS_ADD].is_one() {
+        generate_addition(values);
+    } else if values[IS_SUB].is_one() {
+        generate_subtraction(values);
+    } else if values[IS_MUL].is_one() {
+        generate_multiplication(values);
+    } else if values[IS_DIV].is_one() {
+        generate_division(values);
+    }
+}
+
+pub(crate) fn eval_arithmetic_unit<F: Field, P: PackedField<Scalar = F>>(
+    vars: StarkEvaluationVars<F, P, NUM_COLUMNS, NUM_PUBLIC_INPUTS>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let local_values = &vars.local_values;
+    eval_addition(local_values, yield_constr);
+    eval_subtraction(local_values, yield_constr);
+    eval_multiplication(local_values, yield_constr);
+    eval_division(local_values, yield_constr);
```

Comment:
We also need a constraint of the form `is_add + is_sub + is_mul + is_div = 1`, right?

Reply:
Hmm this makes sense, so we'd have instead a constraint like `sum_x is_opcode_x = 1` to check that there's only one set opcode flag?

Reply:
I was thinking it could be constrained based on the opcode we read from the code segment of memory. I.e. we could split the instruction into bits, parse the opcode (maybe the first 7 bits), and store it in an `opcode` column.

Then we could set e.g. `is_opcode_add = (opcode == OPCODE_ADD)`. We could use the inverse technique to test equality, or maybe construct a `is_opcode_add` filter from the opcode's bits (though we'd need intermediate columns to keep it low-degree). So that should prevent more than one opcode from being active at a time.

I guess we also need to recognize invalid opcodes and revert in that case. I think we can check that with `is_opcode_invalid = 1 - \sum is_opcode_[op]`.

Let me know if you have different ideas about how to approach it though.